### PR TITLE
Add vBalance and supply view functions

### PR DESCRIPTION
**Pull request by charlesjhongc**
- Add vBalance view function for user
- Add total vBalance view function for pool global
- Add `ownershipChange` to prevent flash NFT
- Override NFT `_transfer` function to set `ownershipChange` first

### CODE REVIEW

File: contracts/veLON.sol [hunk 1]
```diff
@@ -285,8 +394,8 @@ contract veLON is IveLON, ERC721, Ownable, ReentrancyGuard {
 
     // Go over weeks to fill history and calculate what the current pool point is
     function _syncGlobalPoints(Point memory poolLastPoint, uint256 _epoch) private returns (Point memory, uint256) {
-        // block slope = dBlock/dTime
-        // If last point is already recorded in this block, slope=0
+        // block decliningRate = dBlock/dTime
+        // If last point is already recorded in this block, decliningRate=0
         // But that's ok because we know the block in such case
         uint256 blockSlope = 0;
```
Comment (zeqing-guo) [inline contracts/veLON.sol:400]: Should we change `blockSlope` to `blockDecliningRate`?

File: contracts/veLON.sol [hunk 2]
```diff
@@ -42,6 +43,114 @@ contract veLON is IveLON, ERC721, Ownable, ReentrancyGuard {
         poolPointHistory[0].ts = block.timestamp;
     }
 
+    function vBalanceOf(uint256 _tokenId) external view override returns (uint256) {
+        return _vBalanceOfAtTime(_tokenId, block.timestamp);
+    }
+
+    function vBalanceOfAtTime(uint256 _tokenId, uint256 _t) external view override returns (uint256) {
+        require(_t <= block.timestamp, "Invalid timestamp");
```
Comment (eric6311) [inline contracts/veLON.sol:51]: Does it need to forbid user querying the past voting power? Perhaps users just want to calculate the voting power by themselves?

Comment (charlesjhongc) [reply contracts/veLON.sol:51]: Actually it avoid `_t` is in the future. Historical vBalance querying is available.

File: contracts/veLON.sol (hunk 1 above)
Comment (charlesjhongc) [reply contracts/veLON.sol]: Right, renamed as `blockRate` in 6bbc955

File: contracts/veLON.sol (hunk 2 above)
Comment (eric6311) [reply contracts/veLON.sol:51]: Oh sorry, I misunderstand the symbol haha

**Review by eric6311 — APPROVED** [review]

**Pull request merged**